clang/include/clang/Basic/Attr.td

```diff
@@ -1281,6 +1281,23 @@ def SYCLSpecialClass: InheritableAttr {
   let Documentation = [SYCLSpecialClassDocs];
 }
 
+def SYCLType: InheritableAttr {
+  let Spellings = [Clang<"sycl_type">];
+  let Subjects = SubjectList<[CXXRecord, Enum], ErrorDiag>;
+  let LangOpts = [SYCLIsDevice, SYCLIsHost];
+  let Args = [EnumArgument<"Type", "SYCLType",
+                           ["accessor", "local_accessor", "spec_constant",
```

Comment:
I discussed this with @steffenlarsen prior to implementing this PR and we decided it makes sense to keep `sycl_special_class ` since it is already well documented and used for the specific case of handling types with __init. I don't have a strong opinion on this and am open to removing it if that is what the majority of reviewers prefer.

Reply:
I would like to see this unified, but I won't insist that it be done as part of this PR.

Reply:
> I won't insist that it be done as part of this PR.

Yeah, same for me.

Reply:
@steffenlarsen @bader @zahiraam are you alright with this? Is it ok if we remove `sycl_special_class ` after we've upstreamed and documented it?

Reply:
Oh, I didn't realize it was upstreamed already.

Reply:
It was checked into clang earlier this year - https://reviews.llvm.org/D114483

Reply:
I am of two minds. On one hand, it is definitely cleaner to have it this way, given you would most likely want `sycl_type` in all cases where you have `sycl_special_class`. However, a big drawback is that we can no longer require the `__init` member function on all classes with the attribute as `sycl_type` would be used on enums and classes that are not considered "SYCL special classes".

A solution could be that we require `__init` based on the value in `sycl_type`, but that adds another step to adding a new SYCL type that needs recognized, and if someone forgets to make a valid `__init` function what are the odds they will remember to add that class to the list of SYCL types that need it?

Reply:
> @steffenlarsen @bader @zahiraam are you alright with this? Is it ok if we remove `sycl_special_class ` after we've upstreamed and documented it?

This patch makes the code much cleaner. But I am worried that there is no mandatory __init method for this new attribute. If we take @steffenlarsen's suggestion are we not back to specifying the type class?

Reply:
Right. IIRC community wanted an attribute which could be used to specify types which required special handling (i.e. requires __init). 

> A solution could be that we require __init based on the value in sycl_type, but that adds another step to adding a new SYCL type that needs recognized, and if someone forgets to make a valid __init function what are the odds they will remember to add that class to the list of SYCL types that need it?

Yea right now, all that is required to enable special handling is define the class with sycl_special_class attribute. Removing the attribute will require developers to modify the frontend as well. 

The more I think about it the more I'm leaning towards leaving sycl_special_class as-is. It isn't uncommon for multiple attributes to be specified on classes

Reply:
I spoke to @premanandrao and @Fznamznon via email and we have decided to keep sycl_special_class as-is for now. We can revisit the attribute in the future if need be.